### ISSUE TITLE
Allow building static libraries

### ISSUE THREAD

**Issue by nolange** [NilsBrause/waylandpp#60]
Hello,

this library is not in debian repositories, and I also prefer not having C++ DSOs if possible (ABI-stability & name-mangling).
Two reasons to use static libraries, and I attached a patch to make this a compiletime option. With this patch, I was able to compile the libs and use them for kodi.

I would help you out getting there, but I have some questions regarding the decisions/implementation.

- you use set instead of option (`option(BUILD_SHARED_LIBS "Build shared library" OFF)`)
- pkg-config file would need to somehow be able to display the dependencies for linking DSOs/Executables (`-lwayland-cursor -lwayland-client -lwayland-egl`)
- CMake configs would make much of this easier (if used from CMake) than pkg-config. Any reason (apart from obvious stuff like time) you haven't done this?

the patch would be:

```patch
diff --git a/CMakeLists.txt b/CMakeLists.txt
index eb9b81e..406acb7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -41,6 +41,7 @@ set(BUILD_SCANNER ON CACHE BOOL "whether to build wayland-scanner++")
 set(BUILD_LIBRARIES ON CACHE BOOL "whether to build the libraries")
 set(BUILD_DOCUMENTATION ${DOXYGEN_FOUND} CACHE BOOL "Create and install the HTML based API documentation (requires Doxygen)")
 set(BUILD_EXAMPLES OFF CACHE BOOL "whether to build the examples (requires BUILD_LIBRARIES to be ON)")
+set(BUILD_SHARED_LIBS ON CACHE BOOL "Build shared libraries")
 
 # variables for .pc.in files
 set(prefix "${CMAKE_INSTALL_PREFIX}")
@@ -126,10 +127,17 @@ if(BUILD_LIBRARIES)
 
   # library building helper functions
   function(define_library TARGET CFLAGS LIBRARIES HEADERS)
-    add_library("${TARGET}" SHARED ${ARGN})
+    add_library("${TARGET}" ${ARGN})
+    get_target_property(libtype "${TARGET}" TYPE)
     target_include_directories("${TARGET}" PUBLIC "include" "${CMAKE_CURRENT_BINARY_DIR}") # latter for wayland-version.hpp
     target_compile_options("${TARGET}" PUBLIC ${CFLAGS})
-    target_link_libraries("${TARGET}" ${LIBRARIES})
+    set(LINK_DEPENDENCIES)
+    if(libtype STREQUAL "STATIC_LIBRARY")
+        # TODO: somehow add information to the pkg-config files
+        # set(LINK_DEPENDENCIES " -l${LIBRARIES}")
+    else()
+      target_link_libraries("${TARGET}" PRIVATE ${LIBRARIES})
+    endif()
     set_target_properties("${TARGET}" PROPERTIES
       PUBLIC_HEADER "${HEADERS}"
       VERSION "${PROJECT_VERSION}"
@@ -158,6 +166,7 @@ if(BUILD_LIBRARIES)
   install(FILES ${PROTO_XMLS} ${PROTO_XMLS_EXTRA} ${PROTO_XMLS_UNSTABLE} DESTINATION "${INSTALL_FULL_PKGDATADIR}/protocols")
   install(TARGETS wayland-client++ wayland-client-extra++ wayland-egl++ wayland-cursor++
     LIBRARY DESTINATION "${CMAKE_INSTALL_FULL_LIBDIR}"
+    ARCHIVE DESTINATION "${CMAKE_INSTALL_FULL_LIBDIR}"
     PUBLIC_HEADER DESTINATION "${CMAKE_INSTALL_FULL_INCLUDEDIR}"
     RESOURCE DESTINATION "${INSTALL_FULL_PKGCONFIGDIR}")
 endif()
diff --git a/wayland-client++.pc.in b/wayland-client++.pc.in
index 8baf088..a923c99 100644
--- a/wayland-client++.pc.in
+++ b/wayland-client++.pc.in
@@ -34,4 +34,4 @@ Version: @PROJECT_VERSION@
 URL: https://github.com/NilsBrause/waylandpp
 Requires.private: wayland-client
 Cflags: -I${includedir}
-Libs: -L${libdir} -lwayland-client++
+Libs: -L${libdir} -lwayland-client++@LINK_DEPENDENCIES@
diff --git a/wayland-client-extra++.pc.in b/wayland-client-extra++.pc.in
index 1c4584d..94b2faa 100644
--- a/wayland-client-extra++.pc.in
+++ b/wayland-client-extra++.pc.in
@@ -34,4 +34,4 @@ Version: @PROJECT_VERSION@
 URL: https://github.com/NilsBrause/waylandpp
 Requires: wayland-client++
 Cflags: -I${includedir}
-Libs: -L${libdir} -lwayland-client-extra++
+Libs: -L${libdir} -lwayland-client-extra++@LINK_DEPENDENCIES@
diff --git a/wayland-client-unstable++.pc.in b/wayland-client-unstable++.pc.in
index c5c79b6..708d40d 100644
--- a/wayland-client-unstable++.pc.in
+++ b/wayland-client-unstable++.pc.in
@@ -34,4 +34,4 @@ Version: @PROJECT_VERSION@
 URL: https://github.com/NilsBrause/waylandpp
 Requires: wayland-client++
 Cflags: -I${includedir}
-Libs: -L${libdir} -lwayland-client-unstable++
+Libs: -L${libdir} -lwayland-client-unstable++@LINK_DEPENDENCIES@
diff --git a/wayland-cursor++.pc.in b/wayland-cursor++.pc.in
index ea174f7..e0bed00 100644
--- a/wayland-cursor++.pc.in
+++ b/wayland-cursor++.pc.in
@@ -34,4 +34,4 @@ Version: @PROJECT_VERSION@
 URL: https://github.com/NilsBrause/waylandpp
 Requires.private: wayland-cursor wayland-client++
 Cflags: -I${includedir}
-Libs: -L${libdir} -lwayland-cursor++
+Libs: -L${libdir} -lwayland-cursor++@LINK_DEPENDENCIES@
diff --git a/wayland-egl++.pc.in b/wayland-egl++.pc.in
index 9afbd4b..2cecd61 100644
--- a/wayland-egl++.pc.in
+++ b/wayland-egl++.pc.in
@@ -34,4 +34,4 @@ Version: @PROJECT_VERSION@
 URL: https://github.com/NilsBrause/waylandpp
 Requires.private: egl wayland-egl wayland-client++
 Cflags: -I${includedir}
-Libs: -L${libdir} -lwayland-egl++
+Libs: -L${libdir} -lwayland-egl++@LINK_DEPENDENCIES@
```

usable like this (added LTO for bragging rights):
```shell
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INTERPROCEDURAL_OPTIMIZATION=ON -DCMAKE_POLICY_DEFAULT_CMP0069=NEW -DBUILD_SHARED_LIBS=OFF  -DBUILD_DOCUMENTATION=OFF -DBUILD_LIBRARIES=ON -DBUILD_EXAMPLES=OFF  ../waylandpp
```

**Comment by NilsBrause** [comment]
Hi!
Sorry for the late answer. I was on holiday.
I needed to remove the PRIVATE keyword from the target_link_libraries line. Otherwise, I get the following errors:

> The keyword signature for target_link_libraries has already been used with
> the target "wayland-client++".  All uses of target_link_libraries
> with a target must be either all-keyword or all-plain.

For setting LINK_DEPENDENCIES, I came up with the following:

```
foreach(LIB ${LIBRARIES})
  set(LINK_DEPENDENCIES "${LINK_DEPENDENCIES} -l${LIB}")
endforeach()
```

This seems to generate correct *.pc files. Does this work for you? If it does, feel free to send a pull request. 
As for "option()" vs "set()", I didn't know about "option()" back then. But I agree that "option()" is probably better. Feel free to send a pull request.
As for CMake configs, I'm using pkg-config, since this can be used by every build system and not only CMake. But if you want to contribute CMake configs, feel free to send a pull request.
